sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/LogicalTypes.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.schemas;
+
+import static org.apache.beam.vendor.guava.v20_0.com.google.common.base.Preconditions.checkArgument;
+
+import java.util.Arrays;
+import org.apache.beam.sdk.schemas.Schema.FieldType;
+import org.apache.beam.sdk.schemas.Schema.LogicalType;
+
+/** A collection of common {@link Schema.LogicalType} classes. */
+public class LogicalTypes {
+  /** A base class for LogicalTypes that use the same Java type as the underlying base type. */
+  public abstract static class PassThroughLogicalType<T> implements LogicalType<T, T> {
+    private final String identifier;
+    private final FieldType fieldType;
+
+    protected PassThroughLogicalType(String identifier, FieldType fieldType) {
+      this.identifier = identifier;
+      this.fieldType = fieldType;
+    }
+
+    @Override
+    public String getIdentifier() {
+      return identifier;
+    }
+
+    @Override
+    public FieldType getBaseType() {
+      return fieldType;
+    }
+
+    @Override
+    public T toBaseType(T input) {
+      return input;
+    }
+
+    @Override
+    public T toInputType(T base) {
+      return base;
+    }
+  }
+
+  /** A LogicalType representing a fixed-size byte array. */
+  public static class FixedBytes implements LogicalType<byte[], byte[]> {
+    public static final String IDENTIFIER = "FixedBytes";
+    private final int byteArraySize;
+
+    private FixedBytes(int byteArraySize) {
+      this.byteArraySize = byteArraySize;
+    }
+
+    public static FixedBytes of(int byteArraySize) {
+      return new FixedBytes(byteArraySize);
+    }
+
+    public int getLength() {
+      return byteArraySize;
+    }
+
+    @Override
+    public String getIdentifier() {
+      return IDENTIFIER;
+    }
+
+    @Override
+    public String getArgument() {
+      return Integer.toString(byteArraySize);
+    }
+
+    @Override
+    public FieldType getBaseType() {
+      return FieldType.BYTES;
+    }
+
+    @Override
+    public byte[] toBaseType(byte[] input) {
+      checkArgument(input.length == byteArraySize);
+      return input;
+    }
+
+    @Override
+    public byte[] toInputType(byte[] base) {
```

Comment:
I am not quite understand the intention of `toBaseType` and `toInputType` of LogicalType, especially when I see this implantation allows a partial copy while  `toBaseType` checks equality on `input.length` and `byteArraySize`.

Why `toInputTypes` allows `base.length <= byteArraySize` but not strictly check `base.length == byteArraySize`?

Reply:
So, the main point is to actually convert types. For example, imagine if we got rid of DATETIME as a primitive schema type and replaced it with a LogicalType (which we may do). In this case the underlying Beam type would be INT64, but we want users to be able to use Joda DateType objects. In that case, the logical type would have the following signature:

class DateTimeType implements LogicalType<DateTime, Long>

But since the user is passing in DateTime objects, we need to know how to convert them back and forth to the Long object that Row expects for an INT64. So you would also override:

Long toBaseType(DateTime dateTime) {
   return dateTime.toMillis();
}

DateTime toInputType(Long millis) {
  return new DateTime(millis);
}

In this particular case we use toBaseType to verify the input, since the java types are the same. However we also allow passing in an array that is smaller than the fixed size, in which case we extend it (possibly we should do this in both codepaths though).

Reply:
Gotcha. Thanks for explanation. So this two functions are util functions for conversions.